src/main/haskell/language-kore/src/Data/Kore/ASTUtils/Substitution.hs

```diff
@@ -0,0 +1,127 @@
+{-|
+Module      : Data.Kore.ASTUtils.Substitution
+Description : Substitute phi_1 for phi_2, avoiding capture
+              In particular this implements axiom 7 in
+              the "large" axiom set (Rosu 2017).
+Copyright   : (c) Runtime Verification, 2018
+License     : UIUC/NCSA
+Maintainer  : phillip.harris@runtimeverification.com
+Stability   : experimental
+Portability : portable
+-}
+
+{-# LANGUAGE MultiParamTypeClasses  #-}
+{-# LANGUAGE FunctionalDependencies #-}
+{-# LANGUAGE Rank2Types             #-}
+{-# LANGUAGE AllowAmbiguousTypes    #-}
+{-# LANGUAGE FlexibleContexts       #-}
+{-# LANGUAGE BangPatterns           #-}
+{-# LANGUAGE LambdaCase             #-}
+{-# LANGUAGE DeriveFunctor          #-}
+{-# LANGUAGE DeriveFoldable         #-}
+{-# LANGUAGE DeriveTraversable      #-}
+{-# LANGUAGE TypeFamilies           #-}
+{-# LANGUAGE PatternSynonyms        #-}
+{-# LANGUAGE ScopedTypeVariables    #-}
+{-# LANGUAGE TypeApplications       #-}
+{-# OPTIONS_GHC -Wno-unused-matches #-}
+{-# OPTIONS_GHC -Wno-missing-signatures #-}
+{-# OPTIONS_GHC -Wno-name-shadowing #-}
+
+module Data.Kore.ASTUtils.Substitution
+( subst 
+, localSubst
+)
+where
+
+
+import           Control.Lens
+
+import qualified Data.Set as S
+import           Data.Fix
+
+import           Data.Kore.AST.Common
+import           Data.Kore.AST.MetaOrObject
+import           Data.Kore.AST.PureML
+
+import           Data.Kore.ASTUtils.SmartConstructors
+
+-- | subst phi_1 phi_2 phi = phi[phi_2/phi_1]
+-- Note that different papers use different conventions.
+-- Here `phi_1` is the old pattern that disappears
+-- and `phi_2` is the new pattern that replaces it. 
+subst
+    :: MetaOrObject level 
+    => CommonPurePattern level 
+    -> CommonPurePattern level 
+    -> CommonPurePattern level 
+    -> CommonPurePattern level
+subst old new = \case 
+    Forall_ s1 v p -> handleBinder old new Forall_ s1 v p 
+    Exists_ s1 v p -> handleBinder old new Exists_ s1 v p 
+    pat 
+     | pat == old -> new
+     | otherwise  -> Fix $ fmap (subst old new) $ unFix pat 
+
+-- subst2 a b = cata (\pat -> if Fix pat == a then b else Fix pat)
+
+-- handleBinder
```

Comment:
You can leave it if it makes sense, but I'm not a fan of code that's commented out. If you allow time to pass, the code will be more and more out of date, but people will avoid deleting it since they assume it has a purpose.